### PR TITLE
[BEAM-4077] Refactor schemas and fields, simplify a bit

### PR DESCRIPTION
**Pull request by kennknowles**
This refactor makes it just a bit easier to construct types and clarifies a bit the difference between TypeName (which is not a type) and FieldType (which is a type) more explicit and makes it easier and quicker to construct types. It also makes the 95% case of non-null types a bit pithier.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @akedin 

I think many of these cleanups are suggestions from you. I think there's more that could be done, too, and I am happy to spend a few more minutes. Like the `withCollectionType` methods and all that are not actually needed any more.

**Comment by kennknowles** [comment]
Ah, the calcite <-> Beam mapping was a bit funky since it mixed up type names and types. I will need to spend a little more time on that.

**Comment by kennknowles** [comment]
OK, I rebuilt the commit history and removed RowSqlTypes. I kept some of it in CalciteUtils. I think we need to have SQL types independent of Beam types unless Beam adopts the standard types. But I think that doesn't fit perfectly - SQL has many types that are like compound types in Beam.

**Comment by kennknowles** [comment]
R: @XuMingmin 

Are you possibly able to give a review?

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/JdbcDriverTest.java [hunk 1]
```diff
@@ -32,11 +32,17 @@
...
 import org.apache.calcite.jdbc.CalciteConnection;
+import org.junit.Before;
 import org.junit.Test;
 
 /** Test for {@link JdbcDriver}. */
 public class JdbcDriverTest {
 
+  @Before
+  public void before() throws Exception {
+    Class.forName("org.apache.beam.sdk.extensions.sql.impl.JdbcDriver");
```
[the remark this answers is not in the document]
Comment (apilloud) [reply sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/JdbcDriverTest.java]: I'd like to understand the environments that require this. What runtime do they have? This should be handled by the service file for java.sql.Driver in Java 7+. How do we test that path?

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/BeamSqlFnExecutorTestBase.java [hunk 2]
```diff
@@ -84,6 +84,11 @@ public static void prepare() {
             .typeSystem(TYPE_FACTORY.getTypeSystem())
             .build();
 
+    try {
+      Class.forName("org.apache.calcite.jdbc.Driver");
```
[the remark this answers is not in the document]
Comment (apilloud) [reply sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/BeamSqlFnExecutorTestBase.java]: Our tests that require this should probably not be using the calcite JDBC driver at all...

**Comment by kennknowles** [comment]
R: @reuvenlax 

This is in a small sequence of #5490, #5497, #5498 about schema follow-ups to make SQL leverage the built-in schema more, with some more streamlined methods and replacing uses of type name with fully specified types.

**Comment by kennknowles** [comment]
Actually let's spread the load & knowledge a bit more.

R: @tgroh ?

**Pull request merged**